Add BLS-on-BN254 scheme for EVM networks

kevincharm opened the pull request:
Many new rollups & networks are coming online, and many of them need good quality onchain randomness! This PR adds a BLS-on-BN254 scheme that is maximally compatible with EVM networks, with signatures on G1.

A test network has been deployed [here](https://drand0.smoketre.es/ce2b88ba52087c5f4c3fd22e92ba78d1389295ac975d2bb95614a11efc532fa0/info).

A [verifier contract](https://basescan.org/address/0x0632a6df853e00e04b8f2f2f825bfc9a5ee9137a) has been deployed to Base. The repo can be found at [fairylabs/anyrand](https://github.com/fairylabs/anyrand). [Example fulfillment tx](https://basescan.org/tx/0xd258c1d9aa9d1a64633783267775b96548b2471c1a3ba8c2fb614decc059a414).

CluEleSsUK commented:
lgtm! Just waiting on the kyber stuff to get reviewed then we can update the `go.mod` accordingly and merge away

AnomalRoil commented on crypto/schemes.go:
```diff
@@ -248,6 +250,44 @@ func NewPedersenBLSUnchainedG1() (cs *Scheme) {
+	var IdentityHashFunc = func() hash.Hash { h, _ := blake2b.New256(nil); return h }
+	// Unchained means we're only hashing the round number
+	var DigestFunc = func(b hashableBeacon) []byte {
+		h := sha3.NewLegacyKeccak256()
```
Good idea sticking to opcode hashes.

AnomalRoil commented on go.mod:
```diff
@@ -2,6 +2,8 @@ module github.com/drand/drand
 
 go 1.20
 
+replace github.com/drand/kyber v1.2.0 => github.com/kevincharm/kyber v1.2.0-bn254.ea10cf74
+
```
This needs to go, so yeah this PR needs to wait on the new Kyber release.

AnomalRoil commented on test/default.toml:
Curious to know what caused these changes? Did a linter complain?

kevincharm replied:
Yeah I was fiddling with these settings to test the new scheme locally, but my editor autoformatted the file on save. I'll revert it.

kevincharm commented on core/util_test.go:
```diff
@@ -222,7 +222,8 @@ func NewDrandTestScenario(t *testing.T, n, thr int, period time.Duration, beacon
 	dt.thr = thr
 	dt.period = period
-	dt.clock = clock.NewFakeClock()
+	// See: https://github.com/jonboulle/clockwork/commit/276013b7b35d157f1a3e88c12ba6cf7480f8669f
+	dt.clock = clock.NewFakeClockAt(time.Date(1984, time.April, 4, 0, 0, 0, 0, time.UTC))
```
Yeah this is very weird indeed. I first changed it to the newer value before I realised the test had the old value hardcoded.

CluEleSsUK replied:
this was previously buried in the clockworks library - hence why we had a unix epoch literal in the test before

AnomalRoil commented:
So. The only thing that's currently preventing me from merging this is that we're actually doing most of the recent dev on the branch `feature/drand-v2.0.0` would you mind rebasing this against that branch and doing the PR against that branch rather @kevincharm ?

kevincharm commented:
will do!

AnomalRoil commented on crypto/schemes.go:
```diff
@@ -248,6 +250,44 @@ func NewPedersenBLSUnchainedG1() (cs *Scheme) {
+// be optimally compatible with the EVM.
+func NewPedersenBLSBN254UnchainedOnG1Scheme() (cs *Scheme) {
+	var Pairing = bn254.NewSuite()
+	Pairing.SetDomainG1([]byte("BLS_SIG_BN254G1_XMD:KECCAK-256_SSWU_RO_NUL_"))
+
```
Is there really no need to set the G2 domain too?

kevincharm replied:
Since the DST is only used for hashing messages, the G2 DST would never be used for this particular scheme (and the BN254 code in Kyber only supports mapping to G1). Should I add it anyway?

AnomalRoil replied:
sg

kevincharm commented on common/chain/info_test.go:
```diff
@@ -96,7 +99,9 @@ func TestChainInfo(t *testing.T) {
 	data[23] = 0x41
 	_, err = InfoFromJSON(bytes.NewReader(data))
-	require.ErrorContains(t, err, "point is not on")
+	if !strings.Contains(err.Error(), "point is not on") && !strings.Contains(err.Error(), "malformed point") {
```
This is also another slight difference in error messages.

kevincharm commented on common/chain/info_test.go:
```diff
@@ -72,7 +73,6 @@ func TestChainInfo(t *testing.T) {
 	c13, err := InfoFromJSON(&c1Buff)
 	require.NoError(t, err)
 	require.NotNil(t, c13)
-	require.Equal(t, c1, c13)
```
`require.Equal` if I understand correctly, seems to be comparing the values of the underlying points, which doesn't work for the current bn254 implementation, as the public key is transformed into affine coordinates when being marshaled. So when it is unmarshaled back from JSON, the points may differ. I noticed that in the bls12-381 library, point operations always transform points into affine coordinates.

kevincharm commented:
@AnomalRoil I found a couple of issues running the tests under the new scheme (please see the inline comments). Sorry I should have double checked the drand tests when working on kyber, some of these could have been avoided.

Also I cannot get the `TestBeaconSimple` test to pass, keeps running into a race condition. Any ideas on what could be the issue there?

kevincharm commented:
I opened a new PR on kyber to address the race issues and get the drand tests passing: https://github.com/drand/kyber/pull/59

kevincharm commented:
I updated kyber to the latest commit (no tag yet), and all tests appear to be passing on my machine with SCHEME_ID=bls-bn254-unchained-on-g1. Also added this scheme to the rest of the other CI test jobs, so hopefully everything passes there 🤞
